### PR TITLE
fix: stop __version__ drift (#85) and mypy [no-redef] in optional mcp import (#87)

### LINKED ISSUE #85: agent_kernel.__version__ reports "0.5.0" but pyproject.toml declares 0.7.0
## Type
Bug

## Priority
P1

## Newcomer impact
Anyone calling `agent_kernel.__version__` for diagnostics, bug reports, or compatibility logic will read the wrong version. Two releases (0.6.0 and 0.7.0) shipped without updating the module-level constant.

## Evidence
- File/path: `src/agent_kernel/__init__.py:105` → `__version__ = "0.5.0"`.
- File/path: `pyproject.toml:7` → `version = "0.7.0"`.
- `CHANGELOG.md` confirms `[0.6.0] - 2026-05-19` and `[0.7.0] - 2026-05-20` were released.
- `pip show weaver-kernel` reports `0.7.0`.
- Command/action tried: `python -c "import agent_kernel; print(agent_kernel.__version__)"`
- Actual behavior: prints `0.5.0`.
- Expected behavior: prints `0.7.0` (matching pyproject + installed dist).

## Steps to reproduce
1. `pip install weaver-kernel==0.7.0`
2. `python -c "import agent_kernel; print(agent_kernel.__version__)"`
3. Output: `0.5.0` (wrong).

## Proposed fix
Two options — implementation agent should pick one and stick with it:

**Option A (recommended — single source of truth via `importlib.metadata`):**
```python
# src/agent_kernel/__init__.py
from importlib.metadata import version as _pkg_version, PackageNotFoundError
try:
    __version__ = _pkg_version("weaver-kernel")
except PackageNotFoundError:  # editable install or unreleased
    __version__ = "0.0.0+local"
```
Eliminates the drift entirely, costs ~3 lines, no runtime cost beyond first-import.

**Option B (cheap and equivalent for now):**
```python
__version__ = "0.7.0"
```
…and add a `RELEASE.md` checklist item / pre-commit hook that bumps both pyproject and `__init__.py` together.

## Implementation notes for AI
- Likely files to inspect/change: `src/agent_kernel/__init__.py`. If Option B, also update `RELEASE.md`.
- Relevant tests to add or update: a new `tests/test_version.py` (or addition to existing) that asserts `__version__ == importlib.metadata.version("weaver-kernel")`.
- Edge cases: the package name on PyPI (`weaver-kernel`) differs from the import name (`agent_kernel`). The `importlib.metadata.version()` call must use the *distribution* name, not the import name.
- Backward compatibility: none — the constant is currently wrong, not API-breaking.

## Acceptance criteria
- [ ] `agent_kernel.__version__` matches `pyproject.toml`'s `[project].version` for every release.
- [ ] A test enforces the equality (or a release-time CI check does).
- [ ] No reference to the obsolete `"0.5.0"` literal remains.

## Duplicate check
Searched open issues for: `__version__`, `version mismatch`, `pyproject.toml`, `0.5.0`, `0.7.0`.
Result: no open duplicate.

### PR DESCRIPTION
**Pull request by dgenio**
Fixes #85 and #87 — the two remaining infra/correctness defects from the `#85`–`#88` triage batch (the other two, #86 and #88, were already fixed in `d83609d`).

## What changed

**#85 — `agent_kernel.__version__` drifted from `pyproject.toml`**
- `src/agent_kernel/__init__.py`: `__version__` is now derived from installed distribution metadata via `importlib.metadata.version("weaver-kernel")` (single source of truth), with a `PackageNotFoundError` → `"0.0.0+local"` fallback for source trees without dist metadata. The hardcoded `"0.5.0"` literal is gone.
- `tests/test_version.py` (new): asserts `__version__ == importlib.metadata.version("weaver-kernel")` and that `__version__` stays in `__all__`.

**#87 — optional `McpError` import triggered mypy `[no-redef]` when `mcp` was absent**
- `src/agent_kernel/drivers/mcp.py`: replaced the annotate-then-`import ... as _McpError` pattern with a `_load_mcp_error()` helper that imports through `importlib.import_module` (mirroring the existing `mcp_support.import_optional` idiom). `_McpError` is now declared exactly once, so the `[no-redef]` is gone.
- `tests/test_mcp_driver.py`: added two regression tests covering both branches of `_load_mcp_error` (mcp present → returns `McpError`; mcp absent → returns `None`).

**Docs**
- `CHANGELOG.md`: added `### Fixed` entries for #85 and #87 under `[Unreleased]`, matching the existing issue-referenced style.

## Why
- #85: `__version__` reported `0.5.0` while the package shipped `0.7.0`/`0.8.0`. `RELEASE.md` only ever documented bumping `pyproject.toml`, so the literal was a latent drift bug; deriving from metadata aligns the code with the documented release process and makes future drift impossible.
- #87: the old pattern declared `_McpError` twice (annotation + `import as`), which mypy reads as a redefinition once the `mcp` import resolves to `Any` under `ignore_missing_imports` (i.e. on a base install without the `[mcp]` extra).

## How verified
All commands run in a clean dev venv (`pip install -e ".[dev,mcp]"`):

- **Reproduced before fixing:**
  - #85: `metadata: 0.8.0 | module: 0.5.0`
  - #87 (with `mcp` uninstalled): `mcp.py:28: error: Name "_McpError" already defined on line 26 [no-redef]`
- **After fix:**
  - #85: `metadata: 0.8.0 | module: 0.8.0`
  - #87: `python -m mypy src/agent_kernel/drivers/mcp.py` → `Success` **both with and without** the `mcp` extra installed.
- **`make ci`** → exit 0: `ruff format --check` clean, `ruff check` `All checks passed!`, `mypy` `Success: no issues found in 41 source files`, **564 passed / 1 skipped**, coverage 94%, examples ran.
- New tests: `tests/test_version.py` (2) + `_load_mcp_error` cases (2) all pass; the mcp-absent branch is covered.

## Scope notes (Mode B)
Strictly limited to #85 and #87. Diffstat: 5 files, +62/−10 (4 tracked + 1 new test). No back-compat shims (per request — none needed; both are internal/diagnostic surfaces). Other open triage issues (#92–#96, #99, #94) are independent feature/docs work and intentionally not bundled here.

## Risks / caveats
- `importlib.metadata.version()` must use the **distribution** name (`weaver-kernel`), not the import name (`agent_kernel`) — handled and covered by `test_version.py`. In an uninstalled raw source tree the value is `0.0.0+local`; in editable/CI installs it reflects the real dist version.

https://claude.ai/code/session_0185WabrTuoL3jCNZiER7ES6

---
_Generated by [Claude Code](https://claude.ai/code/session_0185WabrTuoL3jCNZiER7ES6)_

### CODE REVIEW

File: src/agent_kernel/drivers/mcp.py [hunk 1]
```diff
@@ -19,15 +20,29 @@
...
+    optional-dependency branch is testable. Resolving the type into a separate,
+    explicitly annotated module global avoids the mypy ``[no-redef]`` that the
+    old ``import ... as _McpError`` pattern produced when ``mcp`` was absent and
+    the import resolved to ``Any`` under ``ignore_missing_imports``.
+    """
+    try:
+        exceptions = importlib.import_module("mcp.shared.exceptions")
+    except ImportError:
+        return None
+    error_type: type[BaseException] = exceptions.McpError
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/agent_kernel/drivers/mcp.py]: Good catch — fixed in ffcb2e2. The attribute access is now guarded: `getattr(exceptions, "McpError", None)` followed by an `isinstance(..., type) and issubclass(..., BaseException)` check, so a module that imports but lacks/renames `McpError` degrades to `None` instead of raising `AttributeError` at import time — restoring the old `ImportError → None` semantics for the optional-dep path. Added a regression test (`test_load_mcp_error_returns_none_when_attribute_missing`) covering the module-present-but-attribute-missing branch. `make ci` is green (565 passed, mypy clean).

---
_Generated by [Claude Code](https://claude.ai/code)_
